README.markdown

```diff
@@ -190,49 +190,73 @@ let colour = "red"
 
 ## Code Organization
 
-Use extensions to organize your code into logical blocks of functionality. Each extension should be set off with a `// MARK: -` comment to keep things well-organized. 
-<!-- [@skarol] - this is a little bit controversial as extensions are not meant to organize code. Moreover, grouping part of codes in extensions usually means that class/struct has too many responsibilities and this should be avoided. However having a model class/struct that has to conform to some protocol might fit here. -->
+Separate functionality into different types to follow the [single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle).
+
+If needed use extensions to organize your code into logical blocks of functionality. Each extension should be set off with a `// MARK: -` comment to keep things well-organized. In general this should not be needed if you follow single-responsibility principle.
+
 
 ### Protocol Conformance
 
-In particular, when adding protocol conformance to a model, prefer adding a separate extension for the protocol methods. This keeps the related methods grouped together with the protocol and can simplify instructions to add a protocol to a class with its associated methods.
-<!-- [@skarol] - this example is bad. I do agree with Goranche - we should separate DataSources and Delegates to separate classes. I know there is a lot of places in code where it is used like this, so we should point a direction here. -->
+When adding protocol conformance to a model, prefer adding a separate extension for the protocol methods. This keeps the related methods grouped together with the protocol and can simplify instructions to add a protocol to a class with its associated methods.
+
+Since the compiler does not allow you to re-declare protocol conformance in a derived class, it is not always required to replicate the extension groups of the base class. This is especially true if the derived class is a terminal class and a small number of methods are being overridden. When to preserve the extension groups is left to the discretion of the author.
+
+For DataSource, Delegate, etc conformances try to separate classes wherever it is possible and keep them out of view controller. If this is not possible put protocol conformances into extensions.
+
+For UIKit view controllers, consider grouping lifecycle, custom accessors, and IBAction with `// MARK:`.
 
 **Preferred**:
 ```swift
-class MyViewController: UIViewController {
-  // class stuff here
+class ViewController: UIViewController {
+  let dataSource: DataSource
+  let scrollDelegate: ScrollDelegate
+  // ...
+}
+
+class DataSource: UICollectionViewDataSource 
+  // ...
 }
 
-// MARK: - UITableViewDataSource
-extension MyViewController: UITableViewDataSource {
-  // table view data source methods
+// MARK: - UICollectionViewDataSource conformance
+extension DataSource: UICollectionViewDataSource {
+  // UICollectionViewDataSource methods
 }
```

Comment:
You have redundant `UICollectionViewDataSource` conformance. 
BTW I think if the main purpose of the class is being a `DataSource` and it has a `DataSource` in the name it should conform to `UICollectionViewDataSource` in the type declaration, not in extension. Same for delegates etc.

Reply:
perhaps we should also explain it above